client/state/utils.js

```diff
@@ -268,3 +268,59 @@ export const withSchemaValidation = ( schema, reducer ) => ( state, action ) =>
 
 	return reducer( state, action );
 };
+
+/**
+ * A function that creates action thunks to for performing side-effects in a generic way.
+ * The created thunk will dispatch [FSA](https://github.com/acdlite/flux-standard-action) compliant actions.
+ *
+ * @example
+ * const requestTeams = ( siteId ) => createActionThunk( {
+ *	dataFetch: () => wpcom.undocumented().me().blockSite( siteId ),
+ *	meta: { siteId },
+ * 	requestAction: 'READER_SITE_BLOCK_REQUEST',
+ * 	successAction: 'READER_SITE_BLOCK_REQUEST_SUCCESS',
+ * 	failureAction: 'READER_SITE_BLOCK_REQUEST_FAILURE',
+ * } );
+ *
+ * @param {Function} - dataFetch a 0-argument function for fetching data
+ * @param {Object} meta - extra data to include in the resulting dispatches
+ * @param {Object} requestAction - the action to dispatch when the request has been initiated
+ * @param {Object} successAction - the action to dispatch if the request succeeds. defaults to requestAction_SUCCESS
+ * @param {Object} failureAction - the action to dispatch if the request fails. defaults to requestAction_FAILURE
+ *
+ * @returns {Function} Action thunk for performing side-effects with redux
+ */
+export const createActionThunk = ( {
+	dataFetch,
+	meta = {},
+	requestAction,
+	successAction = `${ requestAction }_SUCCESS`,
+	failureAction = `${ requestAction }_FAILURE`,
+} ) => dispatch => {
+
+	dispatch( {
+		type: requestAction,
+		meta,
+	} );
+
+	return dataFetch()
+		.then(
+			data => {
+				dispatch( {
+					type: successAction,
+					payload: data,
+					error: false,
+					meta,
+				} );
+			} )
+		.catch(
```

Comment:
should probably be part of the initial `then` call, as the second parameter. Otherwise this will also catch any failures when success is dispatched. I think we want to dispatch one or the other, not both.